### PR TITLE
Add Named Pipe Library 

### PR DESCRIPTION
**Pull request by mikeparker**
* Moved IDial and IListen into core project
* Added NamedPipe library

### CODE REVIEW

File: src/HttpOverStream.NamedPipe/NamedPipeDialer.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        private readonly string _serverName;
+        private readonly PipeDirection _pipeDirection;
+        private readonly PipeOptions _pipeOptions;
+        private readonly int _timeout;
+
+        public NamedPipeDialer(string pipeName) 
+            : this(pipeName, ".", PipeDirection.InOut, PipeOptions.Asynchronous | PipeOptions.WriteThrough, 0)
+        {
+        }
+        public NamedPipeDialer(string pipeName, string serverName, PipeDirection pipeDirection, PipeOptions pipeOptions, int timeout)
```
Comment (mikeparker) [inline src/HttpOverStream.NamedPipe/NamedPipeDialer.cs]: I added this extra constructor for extra flexibility

File: src/HttpOverStream.NamedPipe/NamedPipeListener.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        private readonly PipeDirection _pipeDirection;
+        private readonly PipeOptions _pipeOptions;
+        private readonly PipeTransmissionMode _pipeTransmissionMode;
+        private readonly int _maxAllowedServerInstances;
+
+        public NamedPipeListener(string pipeName)
+            : this(pipeName, PipeDirection.InOut, PipeOptions.Asynchronous | PipeOptions.WriteThrough, PipeTransmissionMode.Byte, 0)
+        {
+        }
+        public NamedPipeListener(string pipeName, PipeDirection pipeDirection, PipeOptions pipeOptions, PipeTransmissionMode pipeTransmissionMode, int maxAllowedServerInstances)
```
Comment (mikeparker) [inline src/HttpOverStream.NamedPipe/NamedPipeListener.cs]: I added the 2nd constructor for more flexibility

File: src/HttpOverStream.sln [hunk 3]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio Version 16
-VisualStudioVersion = 16.0.28606.126
+# Visual Studio 15
+VisualStudioVersion = 15.0.28307.329
```
Comment (mikeparker) [inline src/HttpOverStream.sln:4]: Sorry about the churn!

File: src/HttpOverStream.NamedPipe/NamedPipeDialer.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            _pipeOptions = pipeOptions;
+            _timeout = timeout;
+        }
+
+        public async ValueTask<Stream> DialAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            var pipe = new NamedPipeClientStream(_serverName, _pipeName, _pipeDirection, _pipeOptions);
+            await pipe.ConnectAsync(_timeout, cancellationToken).ConfigureAwait(false);
+            return pipe;
+        }
```
Comment (mikeparker) [inline src/HttpOverStream.NamedPipe/NamedPipeDialer.cs:33]: I'm confused why this function doesnt use the `request` parameter. is that correct?

File: src/HttpOverStream.NamedPipe/NamedPipeListener.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            _pipeName = pipeName;
+            _pipeDirection = pipeDirection;
+            _pipeOptions = pipeOptions;
+            _pipeTransmissionMode = pipeTransmissionMode;
+            _maxAllowedServerInstances = maxAllowedServerInstances;
+        }
+
+        public Task StartAsync(Action<Stream> onConnection, CancellationToken cancellationToken)
+        {
+            // TODO: use the parameter cancellationToken
```
Comment (mikeparker) [inline src/HttpOverStream.NamedPipe/NamedPipeListener.cs]: This was copied across from our codebase (i didnt add the TODO comment).

File: src/HttpOverStream.NamedPipe/NamedPipeListener.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+                {
+                    var srv = new NamedPipeServerStream(_pipeName, _pipeDirection, _maxAllowedServerInstances, _pipeTransmissionMode, _pipeOptions);
+                    await srv.WaitForConnectionAsync(ct);
+                    onConnection(srv);
+                }
+            });
+            return Task.CompletedTask;
+        }
+
+        public async Task StopAsync(CancellationToken cancellationToken)
```
Comment (mikeparker) [inline src/HttpOverStream.NamedPipe/NamedPipeListener.cs:46]: Do we need this method at all? Can't we just let the user wait for the task returned from the StartAsync method and use the cancellationtoken / CTS?

File: src/HttpOverStream.NamedPipe/NamedPipeListener.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+            _listenTask = Task.Run(async () =>
+            {
+                while (!ct.IsCancellationRequested)
+                {
+                    var srv = new NamedPipeServerStream(_pipeName, _pipeDirection, _maxAllowedServerInstances, _pipeTransmissionMode, _pipeOptions);
+                    await srv.WaitForConnectionAsync(ct);
+                    onConnection(srv);
+                }
+            });
+            return Task.CompletedTask;
```
Comment (mikeparker) [inline src/HttpOverStream.NamedPipe/NamedPipeListener.cs:43]: Isnt this a bit ugly? Shouldnt we return the _listenTask instead so the user can decide whether to wait for it etc?

File: src/HttpOverStream.NamedPipe/NamedPipeDialer.cs (hunk 4 above)
Comment (mikeparker) [reply src/HttpOverStream.NamedPipe/NamedPipeDialer.cs:33]: Looks like the DialMessageHandler handles this request anyway so should we remove this paramter?
https://github.com/simonferquel/HttpOverStream/blob/master/src/HttpOverStream.Client/DialMessageHandler.cs#L65

File: src/HttpOverStream.NamedPipe/NamedPipeDialer.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    public class NamedPipeDialer : IDial
+    {
+        private readonly string _pipeName;
+        private readonly string _serverName;
+        private readonly PipeDirection _pipeDirection;
+        private readonly PipeOptions _pipeOptions;
+        private readonly int _timeout;
+
+        public NamedPipeDialer(string pipeName) 
+            : this(pipeName, ".", PipeDirection.InOut, PipeOptions.Asynchronous | PipeOptions.WriteThrough, 0)
```
Comment (simonferquel) [inline src/HttpOverStream.NamedPipe/NamedPipeDialer.cs]: WriteThrough is not need anymore

File: src/HttpOverStream.NamedPipe/NamedPipeDialer.cs (hunk 1 above)
Comment (simonferquel) [reply src/HttpOverStream.NamedPipe/NamedPipeDialer.cs]: direction should always be InOut (for http). Adding ImpersonationLevel would be really useful though.

File: src/HttpOverStream.NamedPipe/NamedPipeDialer.cs (hunk 4 above)
Comment (simonferquel) [reply src/HttpOverStream.NamedPipe/NamedPipeDialer.cs:33]: No, it gives useful information to the dialer. (in this particular cas we always call the same named pipe, but other dialers might want to extract the hostname from the request URI or use different logic for GET/POST, etc.)

File: src/HttpOverStream.NamedPipe/NamedPipeListener.cs [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        private Task _listenTask;
+        private CancellationTokenSource _listenTcs;
+        private readonly string _pipeName;
+        private readonly PipeDirection _pipeDirection;
+        private readonly PipeOptions _pipeOptions;
+        private readonly PipeTransmissionMode _pipeTransmissionMode;
+        private readonly int _maxAllowedServerInstances;
+
+        public NamedPipeListener(string pipeName)
+            : this(pipeName, PipeDirection.InOut, PipeOptions.Asynchronous | PipeOptions.WriteThrough, PipeTransmissionMode.Byte, 0)
```
Comment (simonferquel) [inline src/HttpOverStream.NamedPipe/NamedPipeListener.cs]: WriteThrough is not required

File: src/HttpOverStream.NamedPipe/NamedPipeListener.cs (hunk 5 above)
Comment (simonferquel) [reply src/HttpOverStream.NamedPipe/NamedPipeListener.cs]: Actually the TODO is not usefull anymore (we already handle cancellation)

File: src/HttpOverStream.NamedPipe/NamedPipeListener.cs (hunk 7 above)
Comment (simonferquel) [reply src/HttpOverStream.NamedPipe/NamedPipeListener.cs:43]: No, the caller is actually the Owin/Asp.net core Host. This starting method is async for allowing more complex Startup to occur but in the case of NamedPipeListener, the Start implementation is actually synchronous.

One can stop and wait the listener to complete using StopAsync bellow.

File: src/HttpOverStream.NamedPipe/NamedPipeListener.cs (hunk 6 above)
Comment (simonferquel) [reply src/HttpOverStream.NamedPipe/NamedPipeListener.cs:46]: See previous comment

File: src/HttpOverStream.sln (hunk 3 above)
Comment (simonferquel) [reply src/HttpOverStream.sln:4]: No problem, this will just prompt me for updating projects next time I open the solution :D

**Review by simonferquel — CHANGES_REQUESTED** [review]
Few nits, otherwise LGTM

File: src/HttpOverStream.NamedPipe/NamedPipeListener.cs (hunk 9 above)
Comment (mikeparker) [reply src/HttpOverStream.NamedPipe/NamedPipeListener.cs]: 👍

File: src/HttpOverStream.NamedPipe/NamedPipeListener.cs (hunk 5 above)
Comment (mikeparker) [reply src/HttpOverStream.NamedPipe/NamedPipeListener.cs]: Ok so we're happy ignoring the cancellation token passed into both start and stop async? And using our own?

Comment (mikeparker) [reply src/HttpOverStream.NamedPipe/NamedPipeListener.cs]: To clarify, if someone cancels the token passed into StartAsync, it won't do anything here. Is that OK?

Comment (mikeparker) [reply src/HttpOverStream.NamedPipe/NamedPipeListener.cs]: I guess the question is: Whats the point having a cancellation token if we have start/stop methods on the interface anyway? Is someone supposed to cancel the token or call stop? They seem like their purpose overlaps.

File: src/HttpOverStream.NamedPipe/NamedPipeDialer.cs (hunk 1 above)
Comment (mikeparker) [reply src/HttpOverStream.NamedPipe/NamedPipeDialer.cs]: 👍

File: src/HttpOverStream.NamedPipe/NamedPipeDialer.cs (hunk 4 above)
Comment (mikeparker) [reply src/HttpOverStream.NamedPipe/NamedPipeDialer.cs:33]: 👍

**Review by simonferquel — APPROVED** [review]

**Pull request merged**